source/staking/contracts/Staking.sol

```diff
@@ -0,0 +1,159 @@
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/math/Math.sol";
+
+contract Staking is Ownable {
+  using SafeERC20 for IERC20;
+  using SafeMath for uint256;
+  struct Stake {
+    uint256 cliff;
+    uint256 periodLength;
+    uint256 percentPerPeriod;
+    uint256 initialBalance;
+    uint256 currentBalance;
+    uint256 blockNumber;
+  }
+
+  IERC20 public immutable stakingToken;
+  uint256 public cliff;
+  uint256 public periodLength;
+  uint256 public percentPerPeriod;
+  mapping(address => Stake[]) public stakes;
+
+  constructor(
+    IERC20 _stakingToken,
+    uint256 _cliff,
+    uint256 _periodLength,
+    uint256 _percentPerPeriod
+  ) {
+    stakingToken = _stakingToken;
+    cliff = _cliff;
+    periodLength = _periodLength;
+    percentPerPeriod = _percentPerPeriod;
+  }
+
+  function setVestingSchedule(
+    uint256 _cliff,
+    uint256 _periodLength,
+    uint256 _percentPerPeriod
+  ) external onlyOwner {
+    cliff = _cliff;
+    periodLength = _periodLength;
+    percentPerPeriod = _percentPerPeriod;
+  }
+
+  function stake(uint256 amount) external {
+    stakeFor(msg.sender, amount);
+  }
+
+  function stakeFor(address account, uint256 amount) public {
+    stakes[account].push(
+      Stake(cliff, periodLength, percentPerPeriod, amount, amount, block.number)
+    );
+    stakingToken.safeTransferFrom(account, address(this), amount);
+  }
+
+  function addToStake(uint256 index, uint256 amount) external {
+    addToStakeFor(index, msg.sender, amount);
+  }
+
+  function addToStakeFor(
+    uint256 index,
+    address account,
+    uint256 amount
+  ) public {
+    Stake storage stakeData = stakes[msg.sender][index];
+
+    uint256 newInitialBalance = stakeData.initialBalance.add(amount);
+    uint256 newCurrentBalance = stakeData.currentBalance.add(amount);
+    uint256 newBlockNumber =
+      stakeData.blockNumber +
+        amount.mul(block.number.sub(stakeData.blockNumber)).div(
+          newInitialBalance
+        );
+
+    stakes[msg.sender][index] = Stake(
+      cliff,
+      periodLength,
+      percentPerPeriod,
+      newInitialBalance,
+      newCurrentBalance,
+      newBlockNumber
+    );
+
+    stakingToken.safeTransferFrom(account, address(this), amount);
+  }
+
+  function unstake(uint256 index, uint256 amount) external {
+    Stake storage stakeData = stakes[msg.sender][index];
+    require(
+      block.number.sub(stakeData.blockNumber) >= stakeData.cliff,
+      "cliff not reached"
+    );
+    uint256 withdrawableAmount = availableToUnstake(index, msg.sender);
+    require(amount <= withdrawableAmount, "insufficient claimable amount");
+    stakeData.currentBalance = stakeData.currentBalance.sub(amount);
+    if (stakeData.currentBalance == 0) {
+      // remove stake element if claimable amount goes to 0
+      Stake[] storage accountStakes = stakes[msg.sender];
+      stakeData = accountStakes[accountStakes.length.sub(1)];
```

Comment:
I believe this is not really assigning `accountStakes[accountStakes.length.sub(1)]` to `stakes[msg.sender][index]`

Reply:
will check.

Reply:
fixed in `8af9f667aec62dec50386536093e6a04cf7a5792`